src/Filecoin/Address.cpp

```diff
@@ -16,165 +15,283 @@ namespace TW::Filecoin {
+    }
+
+    const auto data = publicKey.hash({}, Hash::HasherKeccak256, true);
+    Data payload(data.end() - 20, data.end());
+
+    return Address(Type::DELEGATED, ETHEREUM_ADDRESS_MANAGER_ACTOR_ID, payload);
+}
+
+Address Address::delegatedAddress(uint64_t actorID, const Data& payload) {
+    return Address(Type::DELEGATED, actorID, payload);
 }
 
 Address::Address(const std::string& string) {
-    if (!isValid(string))
+    auto addr = Address::fromString(string);
+    if (!addr) {
         throw std::invalid_argument("Invalid address data");
-
-    Type type = parseType(string[1]);
-    // First byte is type
-    bytes.push_back(static_cast<uint8_t>(type));
-    if (type == Type::ID) {
-        uint64_t id = std::stoull(string.substr(2));
-        while (id >= 0x80) {
-            bytes.push_back(((uint8_t)id) | 0x80);
-            id >>= 7;
-        }
-        bytes.push_back((uint8_t)id);
-        return;
     }
 
-    Data decoded;
-    if (!Base32::decode(string.substr(2), decoded, BASE32_ALPHABET_FILECOIN))
+    type = addr->type;
+    actorID = addr->actorID;
+    payload = std::move(addr->payload);
+}
+
+Address::Address(const Data& encoded) {
+    auto addr = Address::fromBytes(encoded);
+    if (!addr) {
         throw std::invalid_argument("Invalid address data");
-    uint8_t payloadSize = Address::payloadSize(type);
+    }
 
-    bytes.insert(bytes.end(), decoded.data(), decoded.data() + payloadSize);
+    type = addr->type;
+    actorID = addr->actorID;
+    payload = std::move(addr->payload);
 }
 
-Address::Address(const Data& data) {
-    if (!isValid(data)) {
+Address::Address(Type type_, uint64_t actorID_, const Data& payload_):
+    type(type_),
+    actorID(actorID_),
+    payload(payload_) {
+    if (!isValidPayloadSize(type_, payload_.size())) {
         throw std::invalid_argument("Invalid address data");
     }
-    bytes = data;
 }
 
-Address::Address(const PublicKey& publicKey) {
-    bytes.push_back(static_cast<uint8_t>(Type::SECP256K1));
-    Data hash = Hash::blake2b(publicKey.bytes, payloadSize(Type::SECP256K1));
-    bytes.insert(bytes.end(), hash.begin(), hash.end());
+Data Address::toBytes() const {
+    return addressToBytes(type, actorID, payload);
 }
 
 std::string Address::string() const {
     std::string s;
     // Main net address prefix
     s.push_back(PREFIX);
     // Address type prefix
-    s.push_back(typeAscii(type()));
-
-    if (type() == Type::ID) {
-        uint64_t id = 0;
-        unsigned shift = 0;
-        for (auto i = 1ul; i < bytes.size(); i++) {
-            if (bytes[i] <= SCHAR_MAX) {
-                id |= static_cast<uint64_t>(bytes[i]) << shift;
-                break;
-            } else {
-                id |= static_cast<uint64_t>(bytes[i] & SCHAR_MAX) << shift;
-                shift += 7;
-            }
-        }
-        s.append(std::to_string(id));
+    s.push_back(typeAscii(type));
```

Comment:
emplace_back

Reply:
`typeAscii(type)` returns `char`, and it can be pushed back as it is. Is `emplace_back` required then?